File: pe011/pe011.py
#!/usr/bin/python

"""
Project Euler, Problem 11:

What is the greatest product of four adjacent numbers in the same direction
(up, down, left, right, or diagonally) in the 20×20 grid (input.txt)?

https://projecteuler.net/problem=11
"""

"""
Implementation note:
The solution could have been more compact and straightforward, but the goal was
to use iterators.
"""

import sys

class IterColumns():
    """
    Iterate over columns in a 2D array; insert 0 on the edge.
    """
    def __init__(self, data):
        self.data = data

    def __iter__(self):
        self.i = -1
        self.j = 0
        return self
    
    def __next__(self):
        self.i += 1
        if self.i == len(self.data):
            return 0
        if self.i > len(self.data):
            self.i = 0
            self.j += 1
        if self.j == len(self.data[0]):
            raise StopIteration
        return self.data[self.i][self.j]

class IterRows():
    """
    Iterate over rows in a 2D array; insert 0 on the edge.
    """
    def __init__(self, data):
        self.data = data

    def __iter__(self):
        self.i = 0
        self.j = -1
        return self
    
    def __next__(self):
        self.j += 1
        if self.j == len(self.data[0]):
            return 0
        if self.j > len(self.data[0]):
            self.i += 1
            self.j = 0
        if self.i == len(self.data):
            raise StopIteration
        return self.data[self.i][self.j]

class IterDiagonals():
    """
    Iterate over primary diagonals (\) in a 2D array; insert 0 on the edge.
    """
    def __init__(self, data):
        self.data = data

    def __iter__(self):
        self.k = -(len(self.data))
        self.new_k_next = True
        return self
    
    def shift_k(self):
        self.k += 1
        if self.k >= len(self.data[0]):
            raise StopIteration
        self.new_k_next = False
        if self.k < 0:
            self.i = -self.k - 1
            self.j = -1
        else:
            self.i = -1
            self.j = self.k - 1
    
    def __next__(self):
        if self.new_k_next:
            self.shift_k()
        self.i += 1
        self.j += 1
        if self.i >= len(data) or self.j >= len(data[0]):
            self.new_k_next = True
            return 0
        return self.data[self.i][self.j]

class IterSDiagonals():
    """
    Iterate over secondary diagonals (/) in a 2D array; insert 0 on the edge.
    """
    def __init__(self, data):
        self.data = data

    def __iter__(self):
        self.k = -(len(self.data))
        self.new_k_next = True
        return self
    
    def shift_k(self):
        self.k += 1
        if self.k >= len(self.data[0]):
            raise StopIteration
        self.new_k_next = False
        if self.k < 0:
            self.i = -self.k - 1
            self.j = len(self.data[0])
        else:
            self.i = -1 
            self.j = len(self.data[0]) - self.k
    
    def __next__(self):
        if self.new_k_next:
            self.shift_k()
        self.i += 1
        self.j -= 1
        if self.i >= len(self.data) or self.j < 0:
            self.new_k_next = True
            return 0
        return self.data[self.i][self.j]

data = []

def read_data(filename):
    try:
        with open(filename, 'r') as fi:
            for l in fi:
                if l.strip() == "":
                    continue
                data.append([ int(x) for x in l.split(' ') ])
            fi.close()
    except IOError:
        sys.stderr.write("Error: failed to read data!\n")
        sys.exit(1)

def get_max_product(iterdata, n):
    """
    Find the greatest product of n adjacent numbers in a 1D array (passed as
    an iterator).
    """
    max_product = 0
    have, product = [], 1
    for value in iterdata:
        if value == 0:
            have, product = [], 1
            continue
        product *= value
        have.append(value)
        if len(have) < n:          
            continue
        if len(have) > n:
            product = int(product / have[-(n + 1)])
        if product > max_product:
            max_product = product
    return max_product

def main():
    read_data("input.txt")
    result = []
    for iterator in [IterRows, IterColumns, IterDiagonals, IterSDiagonals]:
        result.append(get_max_product(iterator(data), 4))
    print(max(result))

if __name__ == "__main__":
    main()

File: pe011/pe011.go
/*
Project Euler, Problem 11:

What is the greatest product of four adjacent numbers in the same direction
(up, down, left, right, or diagonally) in the 20×20 grid (input.txt)?

https://projecteuler.net/problem=11
*/

/*
Implementation note:
The solution could have been more compact and straightforward, but the goal was
to take an iterator approach.
*/

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type iterMode int

const (
	IterOverRow iterMode = iota
	IterOverColumn
	IterOverDiagonal
	IterOverSDiagonal
)

/* Iterator for a 2D array (data) */
type Iter struct {
	data   *[][]int
	i, j   int
	k      int
	onEdge bool
	mode   iterMode
}

func NewIter(data *[][]int) *Iter {
	iter := new(Iter)
	iter.Reset()
	iter.data = data
	return iter
}

func (iter *Iter) Reset() {
	iter.onEdge = false
	if iter.mode == IterOverRow {
		iter.i, iter.j = 0, -1
	} else if iter.mode == IterOverColumn {
		iter.i, iter.j = -1, 0
	} else if iter.mode == IterOverDiagonal {
		iter.k = -len(*iter.data)
		iter.i, iter.j = -iter.k, 0
	} else {
		iter.k = -len(*iter.data)
		iter.i, iter.j = -iter.k, len((*iter.data)[0])
	}
}

func (iter *Iter) Next() bool {
	switch iter.mode {
	case IterOverRow:
		return iter.NextInRow()
	case IterOverColumn:
		return iter.NextInColumn()
	case IterOverDiagonal:
		return iter.NextInDiagonal()
	case IterOverSDiagonal:
		return iter.NextInSDiagonal()
	}
	return false
}

func (iter *Iter) Get() int {
	if iter.onEdge {
		return 0
	}
	return (*iter.data)[iter.i][iter.j]
}

/* Iterate over rows */
func (iter *Iter) NextInRow() bool {
	if iter.onEdge == true {
		iter.onEdge = false
		return true
	}
	iter.j += 1
	if iter.j == len((*iter.data)[0]) {
		iter.j = 0
		iter.i += 1
		iter.onEdge = true
	}
	return iter.i != len(*iter.data)
}

/* Iterate over columns */
func (iter *Iter) NextInColumn() bool {
	if iter.onEdge == true {
		iter.onEdge = false
		return true
	}
	iter.i += 1
	if iter.i == len(*iter.data) {
		iter.i = 0
		iter.j += 1
		iter.onEdge = true
	}
	return iter.j != len((*iter.data)[0])
}

/* Iterate over diagonals */
func (iter *Iter) NextInDiagonal() bool {
	if iter.onEdge == true {
		iter.onEdge = false
		return true
	}
	iter.i += 1
	iter.j += 1
	if iter.i >= len(*iter.data) || iter.j >= len((*iter.data)[0]) {
		iter.k += 1
		if iter.k < 0 {
			iter.i = -iter.k
			iter.j = 0
		} else {
			iter.i = 0
			iter.j = iter.k
		}
		iter.onEdge = true
	}
	return iter.k < len((*iter.data)[0])
}

/* Iterate over secondary diagonals */
func (iter *Iter) NextInSDiagonal() bool {
	if iter.onEdge == true {
		iter.onEdge = false
		return true
	}
	iter.i += 1
	iter.j -= 1
	if iter.i >= len(*iter.data) || iter.j < 0 {
		iter.k += 1
		if iter.k < 0 {
			iter.i = -iter.k
			iter.j = len((*iter.data)[0]) - 1
		} else {
			iter.i = 0
			iter.j = len((*iter.data)[0]) - iter.k - 1
		}
		iter.onEdge = true
	}
	return iter.k < len((*iter.data)[0])
}

func assertNoError(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

func main() {
	const n = 4
	var data [][]int
	f, err := os.Open("input.txt")
	assertNoError(err)
	input := bufio.NewScanner(f)

	for input.Scan() {
		var values []int
		if strings.TrimSpace(input.Text()) == "" {
			continue
		}
		for _, s := range strings.Split(input.Text(), " ") {
			value, err := strconv.Atoi(s)
			assertNoError(err)
			values = append(values, value)
		}
		data = append(data, values)
	}
	f.Close()

	maxProduct := 0
	iterator := NewIter(&data)
	for _, mode := range []iterMode{IterOverRow, IterOverColumn, IterOverDiagonal, IterOverSDiagonal} {
		iterator.mode = mode
		iterator.Reset()

		product := 1
		var have []int
		for iterator.Next() {
			value := iterator.Get()
			if value == 0 {
				have = have[:0]
				product = 1
				continue
			}
			product *= value
			have = append(have, value)
			if len(have) < n {
				continue
			}
			if len(have) > n {
				product = int(product / have[len(have)-n-1])
			}
			if product > maxProduct {
				maxProduct = product
			}
		}
	}
	fmt.Println(maxProduct)
}
